plugins/signals-backend/src/service/router.ts

```diff
@@ -14,44 +14,118 @@
  * limitations under the License.
  */
 import {
+  createLegacyAuthAdapters,
   errorHandler,
   PluginEndpointDiscovery,
 } from '@backstage/backend-common';
 import express, { NextFunction, Request, Response } from 'express';
 import Router from 'express-promise-router';
-import { LoggerService } from '@backstage/backend-plugin-api';
+import {
+  BackstageUserInfo,
+  HttpAuthService,
+  LoggerService,
+  UserInfoService,
+} from '@backstage/backend-plugin-api';
 import * as https from 'https';
 import http, { IncomingMessage } from 'http';
 import { SignalManager } from './SignalManager';
-import {
-  BackstageIdentityResponse,
-  IdentityApi,
-  IdentityApiGetIdentityRequest,
-} from '@backstage/plugin-auth-node';
+import { IdentityApi } from '@backstage/plugin-auth-node';
 import { EventBroker } from '@backstage/plugin-events-node';
 import { WebSocket, WebSocketServer } from 'ws';
+import { Duplex } from 'stream';
 
 /** @public */
 export interface RouterOptions {
   logger: LoggerService;
   eventBroker?: EventBroker;
   identity: IdentityApi;
   discovery: PluginEndpointDiscovery;
+  userInfo: UserInfoService;
+  httpAuth?: HttpAuthService;
 }
 
 /** @public */
 export async function createRouter(
   options: RouterOptions,
 ): Promise<express.Router> {
-  const { logger, identity, discovery } = options;
+  const { logger, userInfo, discovery } = options;
+  const { httpAuth } = createLegacyAuthAdapters(options);
+
   const manager = SignalManager.create(options);
   let subscribedToUpgradeRequests = false;
+  let apiUrl: string | undefined = undefined;
 
   const webSocketServer = new WebSocketServer({
     noServer: true,
     clientTracking: false,
   });
 
+  webSocketServer.on('error', (error: Error) => {
+    logger.error('WebSocket server error', error);
+  });
+
+  webSocketServer.on('close', () => {
+    logger.info('WebSocket server closed');
+  });
+
+  const handleUpgrade = async (
+    request: Request<any, any, any, any, any>,
+    socket: Duplex,
+    head: Buffer,
+  ) => {
+    if (!apiUrl) {
+      apiUrl = await discovery.getBaseUrl('signals');
+    }
+
+    if (!request.url || !apiUrl || !apiUrl.endsWith(request.url)) {
+      return;
+    }
+
+    let userIdentity: BackstageUserInfo | undefined = undefined;
+
+    // Authentication token is passed in Sec-WebSocket-Protocol header as there
+    // is no other way to pass the token with plain websockets
+    try {
+      const token = request.headers['sec-websocket-protocol'];
+      if (token) {
+        const credentials = await httpAuth.credentials(
+          {
+            ...request,
+            headers: { ...request.headers, authorization: `Bearer ${token}` },
+          } as Request<any, any, any, any, any>,
+          { allow: ['user', 'none'], allowLimitedAccess: true },
+        );
+
+        if (credentials.principal.type === 'user') {
```

Comment:
Would it be fine to limit this to only authenticated users at this point? We've got guest access for livable development with proper tokens now thanks to #22565

Reply:
Sure, works for me